### PR TITLE
research: recover local models with PyTorch, ONNX and media E2E

### PR DESCRIPTION
**Pull request by donghaozhang**
## Summary

Local neural-model inventory, bounded PyTorch/ONNX recovery tools, native-reference validation, portable runtime tests, and an explicit ONNX shot-detection option. The existing default engines remain unchanged. Private weights, recovered graphs, alphabets, tensors, and media are not published.

Review entry points:
- `qcut/docs/technical/ai/local-neural-model-inventory-2026-09-19.zh-CN.md`
- `qcut/research/local-model-pytorch/README.zh-CN.md`
- `qcut/research/local-model-pytorch/PHASE5-20260919.zh-CN.md`

## Phase 5 Changes

- Add authored synthetic ONNX platform tests and four-platform CI, separate from private-model evidence.
- Add streaming, bounded no-Torch shot inference with verified model identity, input/output contracts, process cleanup, and an explicit CLI `--engine onnx`.
- Add editor IPC/HTTP routing and an opt-in timeline command, stale-result guards, retry engine preservation, shared platform types, and packaged Python scripts. No weights or Python interpreter are bundled automatically; missing configuration fails explicitly.
- Recover the exact private OCR alphabet and implement bounded CTC decoding and honest image-to-text quality measurements.
- Add a new GRU CPU v3 profile with full-output native parity; register it only through a pinned artifact hash. Reject unverified ONNX export and retain historical candidate failures.
- Add a new Bandou ordered profile, strict independent loader, native comparisons, and bounded ONNX smoke tooling. Shared registration and complete ONNX qualification remain pending.

## Verification

- Final Python discovery: **604/604 passed**.
- Electron/CLI focused Vitest: **109/109 passed** across seven files.
- Frontend context-menu/scene-routing Vitest: **39/39 passed** across two files.
- Electron and frontend TypeScript checks passed. Electron/Vite builds also completed during integration.
- Public ONNX CI [35427584529](https://github.com/Quriosity-agent/qcut/actions/runs/35427584529): Windows x64, Linux x64, Linux ARM64, macOS ARM64 all passed at `6786b918f`. Each ran 33 unit tests plus five authored graphs, 36 numeric cases, 20 reload/reset cases, and five CLI executions. This is not a claim about final-head CI or private Windows models.
- Linux x86_64 private replay: **19 graphs / 76 cases passed**, in an AMD64 container on an ARM host, network disabled, without Torch. Earlier macOS/Linux ARM64 evidence is retained in Phase 4.
- Six real-video/fps shot comparisons matched frame/cut results against the existing Torch path. A no-Torch 60-second **looped fixture** produced 39 cuts in 11.685 seconds. Probability values are not bit-exact; the known preprocessing difference is documented.
- GRU v3: **250 inferences / 1,000 full outputs** passed the unchanged mixed tolerance; all 250 probability tensors were bit-exact, not all state tensors.
- Bandou: **30/30 native**, **30/30 independent replay**, and **54/54 existing-profile regressions**; ONNX smoke only **4/4**.
- OCR: all 11 fresh native/PT/ONNX full-logit comparisons passed; decoded text agreed across engines, but semantic quality was only **4/11 exact, CER 5.0209%**.
- `git diff --check` passed. All 74 finalization commits were verified against exact allowlisted paths and blobs, with zero multi-file commits. Five platform commits were pushed earlier in this phase; the previous 130 one-file commits are retained.

## Remaining Work / Explicit Failures

- Actual Electron E2E: **2 failed**, `Media not found` after import, before inference. Real split/undo/save/reload/preview/export acceptance is not complete. Private failure screenshots and logs are retained.
- Editor cancellation currently discards results; backend cancellation IPC still needs integration. Windows media/process-tree cancellation has not been exercised end to end.
- Windows private-model/media replay remains unverified. The pinned ORT version has no macOS Intel wheel.
- Bandou shared registration, full ONNX replay and broadcast-compaction unit coverage remain pending. GRU portable ONNX/profile and editor matting are not enabled.
- OCR detector/crop/orientation/color preprocessing and editor OCR remain separate work.
- Older Phase 4 counts (21 network versions, 16 PT packages, 19 ONNX files) remain historical snapshots, not current product completion percentages. New serialization profiles are not counted again as model families.

No accuracy threshold was relaxed. Conversion does not grant redistribution rights to vendor assets. CI cannot recreate private oracle evidence from public files alone. The user requested immediate wrap-up, so further experiments were stopped and incomplete acceptance is documented rather than relabeled as successful.

## Summary by CodeRabbit

- **New Features**
  - Added an optional local ONNX engine for video scene detection and shot splitting.
  - Added ONNX shot-splitting controls to the timeline context menu and command-line workflows.
  - Added progress reporting, cancellation, validation, undo/redo support, and media metadata to ONNX scene detection.
  - Added packaged local model support for ONNX-based processing.
  - Added automatic registration of local desktop videos in project media.

- **Bug Fixes**
  - Improved handling of changed media, invalid detection results, failed processes, and cancelled operations without altering the timeline.
  - Improved local video import handling, including existing media and failed imports.

- **Documentation**
  - Added documentation covering local model support, ONNX validation, reproducibility, testing, and known limitations.

### CODE REVIEW

File: qcut/research/local-model-pytorch/matting_cpu_report.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+              "first_failing_layer": diagnosis["first_failure"], "diagnostic_report": str(trace.resolve()),
+              "diagnostics": diagnostics,
+              "isolated_softmax_on_native_logits": diagnosis["isolated_cpu_softmax_on_native_logits"],
+              "synthetic_softmax_proof": {"report": str(math.resolve()), "status": math_proof["status"], "cases": math_proof["cases"]},
+              "all_declared_outputs_verified": passed, "product_gpu_parity": False,
+              "unchanged_container_load_tested": False, "original_fp16_native_expansion_bit_exact": True,
+              "integration": {"module": "matting_torch", "loader": "load_model(path=..., expected_sha256=..., allow_unverified=True)",
+                              "candidate_opt_in_required": True,
+                              "format": "qcut-private-matting-gru-cpu-v2", "vendor_runtime_required_for_loading": False},
+              "scope": "CPU network only; original arena normalization independently native-verified; complete temporal parity failed; no product/editor GPU or preprocessing claim"}
```
[the remark this answers is not in the document]
Comment (Jareddvw) [reply qcut/research/local-model-pytorch/matting_cpu_report.py]: [Claude Code] Fixed in dc40865 — verified against the current code rather than assumed from `outdated`.

The notice is now split in two: `build_workspace_share_review_email` (`db/services/collab_requests/workspace_share_review_notice.py:74`) does the reads and renders the payload inside the transaction, and `send_workspace_share_review_email` (:138) does the actual send. The built payload is carried out on `ApprovalSideEffects.workspace_share_review_email` and dispatched by `_send_deferred_workspace_share_review_email` (`routes/collab_requests.py:103`).

I checked all three settle routes, and in each one the dispatch sits *after* the `with open_txn() as session_tx:` block has closed, at the same indentation as the `with`:
- approve by request id — `routes/collab_requests.py:732`
- approve by resource — `:943`
- decline — `:1042` (the net-new call; this route previously had no dispatch at all)

Both build sites are inside the transaction but are pure — the rejection branch at `db/services/collab_requests/service.py:3314` and the internal-workspace-share approval branch at `:3504` — so a rollback now discards the payload instead of leaving a sent email behind.
